snapshotter/devmapper/devmapper.go

```diff
@@ -0,0 +1,319 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package devmapper
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/containerd/containerd/log"
+	"github.com/containerd/containerd/mount"
+	"github.com/containerd/containerd/snapshots"
+	"github.com/containerd/containerd/snapshots/storage"
+	"github.com/hashicorp/go-multierror"
+	"github.com/moby/moby/pkg/dmesg"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	metadataFileName = "metadata.db"
+	fsTypeExt4       = "ext4"
+)
+
+// devmapper implements containerd's snapshotter based on Linux device-mapper targets.
+type Snapshotter struct {
+	store  *storage.MetaStore
+	pool   *PoolDevice
+	config *Config
+}
+
+func NewSnapshotter(ctx context.Context, configPath string) (*Snapshotter, error) {
+	log.G(ctx).WithField("cfg_path", configPath).Info("creating devmapper snapshotter")
+
+	config, err := LoadConfig(configPath)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := os.MkdirAll(config.RootPath, 0755); err != nil && !os.IsExist(err) {
+		return nil, errors.Wrapf(err, "failed to create root directory: %s", config.RootPath)
+	}
+
+	store, err := storage.NewMetaStore(filepath.Join(config.RootPath, metadataFileName))
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create metastore")
+	}
+
+	poolDevice, err := NewPoolDevice(ctx, config.PoolName, config.DataDevice, config.MetadataDevice, config.DataBlockSizeSectors)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Snapshotter{
+		store:  store,
+		config: config,
+		pool:   poolDevice,
+	}, nil
+}
+
+func (dm *Snapshotter) Stat(ctx context.Context, key string) (snapshots.Info, error) {
+	log.G(ctx).WithField("key", key).Debug("stat")
+
+	ctx, trans, err := dm.store.TransactionContext(ctx, false)
+	if err != nil {
+		return snapshots.Info{}, err
+	}
+
+	defer trans.Rollback()
```

Comment:
I want to be explicit that `Rollback` is always called.

Reply:
Why on Stat() but not on other operations?

Reply:
Not being familiar with the containerd Transactor interface it wasn't obvious to me why having a transaction object on a non-writable operation would require a rollback but it's listed as part of the [interface definition](https://github.com/containerd/containerd/blob/master/snapshots/storage/metastore.go#L41-L44) in case anyone else reads this and is curious.